libraries/eosiolib/core/eosio/to_key.hpp

```diff
@@ -0,0 +1,313 @@
+#pragma once
+
+#include <deque>
+#include <list>
+#include <map>
+#include <optional>
+#include <set>
+#include <tuple>
+#include <utility>
+#include <variant>
+#include <vector>
+#include <cstring>
+
+#include "datastream.hpp"
+#include <bluegrass/meta/refl.hpp>
+
+namespace eosio {
+
+template<typename T>
+constexpr bool has_bitwise_serialization() {
+   if constexpr (std::is_arithmetic_v<T>) {
+      return true;
+   } else if constexpr (std::is_enum_v<T>) {
+      static_assert(!std::is_convertible_v<T, std::underlying_type_t<T>>, "Serializing unscoped enum");
+      return true;
+   } else {
+      return false;
+   }
+}
+
+template <typename... Ts, typename S>
+void to_key(const std::tuple<Ts...>& obj, datastream<S>& stream);
+
+// to_key defines a conversion from a type to a sequence of bytes whose lexicograpical
+// ordering is the same as the ordering of the original type.
+//
+// For any two objects of type T, a and b:
+//
+// - key(a) < key(b) iff a < b
+// - key(a) is not a prefix of key(b)
+//
+// Overloads of to_key for user-defined types can be found by Koenig lookup.
+//
+// Abieos provides specializations of to_key for the following types
+// - std::string and std::string_view
+// - std::vector, std::list, std::deque
+// - std::tuple
+// - std::array
+// - std::optional
+// - std::variant
+// - Arithmetic types
+// - Scoped enumeration types
+// - Reflected structs
+// - All smart-contract related types defined by abieos
+template <typename T, typename S>
+void to_key(const T& obj, datastream<S>& stream);
+
+template <int i, typename T, typename S>
+void to_key_tuple(const T& obj, datastream<S>& stream) {
+   if constexpr (i < std::tuple_size_v<T>) {
+      to_key(std::get<i>(obj), stream);
+      to_key_tuple<i + 1>(obj, stream);
+   }
+}
+
+template <typename... Ts, typename S>
+void to_key(const std::tuple<Ts...>& obj, datastream<S>& stream) {
+   to_key_tuple<0>(obj, stream);
+}
+
+template <typename T, std::size_t N, typename S>
+void to_key(const std::array<T, N>& obj, datastream<S>& stream) {
+   for (const T& elem : obj) { to_key(elem, stream); }
+}
+
+template <typename T, typename S>
+void to_key_optional(const bool* obj, datastream<S>& stream) {
+   if (obj == nullptr)
+      stream.write('\0');
+   else if (!*obj)
+      stream.write('\1');
+   else {
+      stream.write("\2", 2);
```

Comment:
These look like bugs. Will this write {\2, \0}? And is that what we want?

Reply:
As discussed on the call, this does seem to be a bug that was pulled in from old code. I've brought over any remaining differences to include fixing this.